File: java-kata/src/main/java/com/leetcode/EvaluateReversePolishNotation.java
package com.leetcode;

import java.util.ArrayDeque;

/**
 * <h3><a href="https://leetcode.com/problems/evaluate-reverse-polish-notation/">Evaluate Reverse Polish Notation</a></h3>
 * <p>
 *     You are given an array of strings tokens that represents an arithmetic expression
 *     in a <a href="http://en.wikipedia.org/wiki/Reverse_Polish_notation">Reverse Polish Notation</a>.
 *     Evaluate the expression. Return an integer that represents the value of the expression.
 * </p>
 * <p>
 *     <strong>Note that:</strong>
 *     <ul>
 *         <li>The valid operators are <code>'+'</code>, <code>'-'</code>, <code>'*'</code> and <code>'/'</code>.</li>
 *         <li>Each operand may be an integer or another expression.</li>
 *         <li>The division between two integers always <strong>truncates toward zero</strong>.</li>
 *         <li>There will not be any division by zero.</li>
 *         <li>The input represents a valid arithmetic expression in a reverse polish notation.</li>
 *         <li>The answer and all the intermediate calculations can be represented in a <strong>32-bit</strong> integer.</li>
 *     </ul>
 * </p>
 * <p>
 *     <h6>Constraints:</h6>
 *     <ul>
 *         <li>
 *             <code>1 <= tokens.length <= 10^4</code>
 *         </li>
 *         <li>
 *             <code>tokens[i]</code> is either an operator: <code>'+'</code>, <code>'-'</code>, <code>'*'</code> or
 *             <code>'/'</code> or an integer in the range <code>[-200, 200]</code>
 *         </li>
 *     </ul>
 * </p>
 */
public class EvaluateReversePolishNotation {

    public int evalRPN(String[] tokens) {
        var stack = new ArrayDeque<Integer>();
        for (var token : tokens) {
            switch (token) {
                case "+" -> {
                    var right = stack.removeLast();
                    var left = stack.removeLast();
                    stack.add(left + right);
                }
                case "-" -> {
                    var right = stack.removeLast();
                    var left = stack.removeLast();
                    stack.add(left - right);
                }
                case "*" -> {
                    var right = stack.removeLast();
                    var left = stack.removeLast();
                    stack.add(left * right);
                }
                case "/" -> {
                    var right = stack.removeLast();
                    var left = stack.removeLast();
                    stack.add(left / right);
                }
                default -> stack.add(Integer.parseInt(token));
            }
        }
        return stack.element();
    }
}

File: kotlin-kata/src/main/kotlin/com/leetcode/EvaluateReversePolishNotation.kt
package com.leetcode

import java.util.Stack

/**
 * ### [Evaluate Reverse Polish Notation](https://leetcode.com/problems/evaluate-reverse-polish-notation/)
 *
 * You are given an array of strings tokens that represents an arithmetic expression
 * in a [Reverse Polish Notation](http://en.wikipedia.org/wiki/Reverse_Polish_notation).
 * Evaluate the expression. Return an integer that represents the value of the expression.
 *
 * **Note that:**
 *  - The valid operators are `'+'`, `'-'`, `'*'` and `'/'`.
 *  - Each operand may be an integer or another expression.
 *  - The division between two integers always **truncates toward zero**.
 *  - There will not be any division by zero.
 *  - The input represents a valid arithmetic expression in a reverse polish notation.
 *  - The answer and all the intermediate calculations can be represented in a **32-bit** integer.
 *
 * ###### Constraints:
 *  - `1 <= tokens.length <= 10^4`
 *  - `tokens[i]` is either an operator: `'+'`, `'-'`, `'*'` or `'/'` or an integer in the range `[-200, 200]`
 */
class EvaluateReversePolishNotation {

    fun evalRPN(tokens: Array<String>): Int {
        val stack = Stack<Int>()

        tokens.forEach {
            when (it) {
                "+" -> stack.doPlus()
                "-" -> stack.doMinus()
                "*" -> stack.doMultiply()
                "/" -> stack.doDivide()
                else -> stack.push(it.toInt())
            }
        }

        return stack.peek()
    }

    private inline fun Stack<Int>.doPlus() {
        val right = this.pop()
        val left = this.pop()
        this.push(left + right)
    }

    private inline fun Stack<Int>.doMinus() {
        val right = this.pop()
        val left = this.pop()
        this.push(left - right)
    }

    private inline fun Stack<Int>.doMultiply() {
        val right = this.pop()
        val left = this.pop()
        this.push(left * right)
    }

    private inline fun Stack<Int>.doDivide() {
        val right = this.pop()
        val left = this.pop()
        this.push(left / right)
    }
}
